src/appservice/Api.ts

```diff
@@ -0,0 +1,197 @@
+import request from "request";
+import express from "express";
+import * as bodyParser from "body-parser";
+import { MjolnirManager } from "./MjolnirManager";
+import * as http from "http";
+
+/**
+ * This provides a web api that is designed to power the mjolnir widget https://github.com/matrix-org/mjolnir-widget.
+ */
+export class Api {
+    private httpdConfig: express.Express = express();
+    private httpServer?: http.Server;
+
+    constructor(
+        private homeserver: string,
+        private mjolnirManager: MjolnirManager,
+    ) {}
+
+    private resolveAccessToken(accessToken: string): Promise<string> {
+        return new Promise((resolve, reject) => {
+            request({
+                url: `${this.homeserver}/_matrix/federation/v1/openid/userinfo`,
+                qs: { access_token: accessToken },
+            }, (err, homeserver_response, body) => {
+                if (err) {
+                    reject(null);
+                }
+
+                let response: { sub: string};
+                try {
+                    response = JSON.parse(body);
+                } catch (e) {
+                    reject(null);
```

Comment:
Good spot

Reply:
So there is a reason for this, have documented it.

Reply:
It's now in the function doc, but we return `null` because that's how we report back to the user whether the request is authorized or not.

Reply:
But we don't log anything. I think that deserves a little comment explaining why.

Reply:
Ok, having taken a look at this, there is a slightly bigger problem... I don't believe that this will work for remote users. Meaning that the widget api can't currently work with them.

I have changed this to log more since you're right, and i wasn't thinking about things properly/managed to confuse myself. I will document both of the issues with logging in general and the widget api